fala/apps/adviser/forms.py

```diff
@@ -74,62 +78,99 @@ def clean(self):
-        elif re.search(r"^GY[0-8]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.GUERNSEY
-        elif postcode_no_space[:2] in SCOTTISH_PREFIXES:
+        elif country == "Scotland":
             return Region.SCOTLAND
-        else:
+        elif country == "England" or country == "Wales":
             return Region.ENGLAND_OR_WALES
+        else:
+            self.add_error("postcode", _("This service is only available for England and Wales"))
+            return None
 
     @property
     def current_page(self):
         return self.cleaned_data.get("page", 1)
 
-    def _valid_postcode(self, postcode):
+    def validate_postcode_and_return_country(self, postcode):
         try:
             # Check if the postcode is a valid string before proceeding
             if not isinstance(postcode, str) or not postcode.strip():
                 return False
 
-            # Attempt to find the data using the LAALAA API
-            data = laalaa.find(
-                postcode=postcode.strip(),
-                page=1,
-            )
+            # Retry set-up as per LAALAA API
+            session = requests.Session()
+            retry_strategy = Retry(total=5, backoff_factor=0.1)
+            adapter = HTTPAdapter(max_retries=retry_strategy)
+            session.mount("https://", adapter)
+
+            # Call postcodes.io API using `api.postcodes.io/postcodes?q=` endpoint
+            # This one let's you use partial postcodes and returns the country in the `result`
+            url = settings.POSTCODE_IO_URL + f"{postcode}"
+            response = session.get(url, timeout=5)
+
+            # Check if response was successful
+            if response.status_code != 200:
+                return False
+
+            # If the 'result' key is Null, the postcode is invalid
+            data = response.json()
+            if not data.get("result"):
+                return False
+
+            first_result_in_list = data["result"][0]
+            country = first_result_in_list.get("country")
+            nhs_ha = first_result_in_list.get("nhs_ha")
 
-            # Check if the 'error' key is present in the data
-            return "error" not in data
+            if country and nhs_ha:
+                return country, nhs_ha
+            else:
+                return False
 
-        except laalaa.LaaLaaError:
-            # If there is an exception from LAALAA for example a postcode is not supported by our data. e.g. IM4
-            # Tell the user there was an error, but don't stop the usage of the site.
-            self.add_error("postcode", "%s %s" % (_("Error looking up legal advisers."), _("Please try again later.")))
+        except requests.RequestException:
+            # If there is an exception from postcode.io, tell the user there was an error, but don't stop the usage of the site.
+            self.add_error("postcode", _("Error looking up legal advisers via our API. Please try again later."))
```

Comment:
The old error message is fine. "Error looking up legal advisers" The API means nothing to the end user when they see it pop up on the screen.

Have you tested to see this error response works? If not, try purposely changing the URL to the postcode.io wrong and then calling the search function.

Reply:
Restored the wording.

Yes, did test that the error response works.

- I changed it from this:

`POSTCODE_IO_URL = "https://api.postcodes.io/postcodes?q="`

- to this:

`POSTCODE_IO_URL = "https://api.postcode.io/postcodes?q="`

- and got this:

![Screenshot 2024-10-02 at 09 50 44](https://github.com/user-attachments/assets/7f371686-5ddc-472f-8d52-748f0a9d1d8f)